src/target/riscv/riscv.c

```diff
@@ -598,90 +606,174 @@ static int maybe_add_trigger_t1(struct target *target, struct trigger *trigger)
 	return ERROR_OK;
 }
 
-static int maybe_add_trigger_t2(struct target *target, struct trigger *trigger)
-{
-	int ret;
-	riscv_reg_t tdata1, tdata2;
+struct trigger_request_info {
+	riscv_reg_t tdata1;
+	riscv_reg_t tdata2;
+	riscv_reg_t tdata1_ignore_mask;
+};
 
+static int try_trigger_request_match(struct target *target,
+									 struct trigger *trigger, int trigger_type,
+									 struct trigger_request_info trig_info)
+{
+	int ret = ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
 	RISCV_INFO(r);
 
-	tdata1 = 0;
-	tdata1 = set_field(tdata1, CSR_MCONTROL_TYPE(riscv_xlen(target)), 2);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_DMODE(riscv_xlen(target)), 1);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_ACTION, CSR_MCONTROL_ACTION_DEBUG_MODE);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_M, 1);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_S, !!(r->misa & (1 << ('S' - 'A'))));
-	tdata1 = set_field(tdata1, CSR_MCONTROL_U, !!(r->misa & (1 << ('U' - 'A'))));
-	tdata1 = set_field(tdata1, CSR_MCONTROL_EXECUTE, trigger->execute);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_LOAD, trigger->read);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_STORE, trigger->write);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_ANY & 3);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_ANY >> 2) & 3);
-
-	if (trigger->execute || trigger->length == 1)
-		goto MATCH_EQUAL;
-	if (!can_use_napot_match(trigger, &tdata2))
-		goto MATCH_GE_LT;
-
-	unsigned int idx;
-	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL, false, &idx);
-	if (ret != ERROR_OK)
-		return ret;
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_NAPOT);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_CHAIN, CSR_MCONTROL_CHAIN_DISABLED);
-	ret = set_trigger(target, idx, tdata1, tdata2, CSR_MCONTROL_MASKMAX(riscv_xlen(target)));
-	if (ret != ERROR_OK) {
-		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
-			goto MATCH_GE_LT; /* Fallback to chained MATCH using GT and LE */
-		return ret;
+	for (unsigned int idx = 0;
+		 find_next_trigger(target, trigger_type, false, &idx) == ERROR_OK &&
+		 ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
```

Comment:
I much prefer a simpler loop, where you do these calls inside the loop and break/continue as necessary.

Reply:
Thank you for the suggestion. As I see it there are two approaches to rewriting the loop:

```
for (unsigned int idx = 0; ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE; ++idx) {
    if (find_next_trigger(target, trigger_type, false, &idx) != ERROR_OK)
        break;
    /* loop body */
```

There are several reasons why I do not like this approach:

1. The initialization and mutation of `idx` are in the loop's header, but it is unclear on first look that the loop is finite (loop condition does not check the index).

2. The mutations of `idx` are further separated, so one can assume that the loop checks all `idx` values in order, which is untrue.

The other approach:

```

int ret = ERROR_OK;
for (unsigned int idx = 0; find_next_trigger(target, trigger_type, false, &idx) != ERROR_OK; ++idx) {
    /* loop body */
    if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
        break;
}

```

I find that this is the better method since I only see one small drawback: it is not clear at first glance that the loop is here to find the first trigger that can be used, not to simply iterate over triggers of the same type. It is also somewhat better than the initial version since the initialization of `ret` is more straightforward.

Moreover, in both approaches I fail to see in which way the loop is actually simplified. Minor changes to the condition trade off major changes to the body.

However, if you insist, I'm ready to do as you suggested

Reply:
Your "other approach" looks good to me.

Reply:
Addressed.